main.c

```diff
@@ -524,6 +528,139 @@ init_vt(struct vkcube *vc)
+                                                                       .handleType = VK_EXTERNAL_MEMORY_HANDLE_TYPE_DMA_BUF_BIT_EXT,
+                                                                   },
+                                                                   .format = vc->image_format,
+                                                                   .type = VK_IMAGE_TYPE_2D,
+                                                                   .tiling = VK_IMAGE_TILING_LINEAR,
+                                                                   .usage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT,
+                                                               },
+                                                               &image_format_props);
+      fail_if(res != VK_SUCCESS, "Failed to get DeviceImageFormatProperties.\n");
+
+      VkExternalMemoryFeatureFlags feature_flags = external_image_format_props
+          .externalMemoryProperties
+          .externalMemoryFeatures;
+      fail_if(!(feature_flags & VK_EXTERNAL_MEMORY_FEATURE_IMPORTABLE_BIT), "dma-buf does not support Vulkan import");
+
+      VkImageDrmFormatModifierExplicitCreateInfoEXT mod_create_info = {
+          .sType = VK_STRUCTURE_TYPE_IMAGE_DRM_FORMAT_MODIFIER_EXPLICIT_CREATE_INFO_EXT,
+          .pNext = &(VkExternalMemoryImageCreateInfo){
+              .sType = VK_STRUCTURE_TYPE_EXTERNAL_MEMORY_IMAGE_CREATE_INFO,
+              .handleTypes = VK_EXTERNAL_MEMORY_HANDLE_TYPE_DMA_BUF_BIT_EXT,
+          },
+          .drmFormatModifier = drm_format_mod,
+          .drmFormatModifierPlaneCount = 1,
+          .pPlaneLayouts = (VkSubresourceLayout[]) {
+              {
+                  .offset = gbm_bo_get_offset(b->gbm_bo, 0),
+                  .rowPitch = gbm_bo_get_stride_for_plane(b->gbm_bo, 0),
+              },
+          },
+      };
+
+  VkImageCreateInfo base_create_info = {
+          .sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO,
+          .pNext = &mod_create_info,
+          .imageType = VK_IMAGE_TYPE_2D,
+          .format = vc->image_format,
+          .extent = (VkExtent3D){
+              gbm_bo_get_width(b->gbm_bo),
+              gbm_bo_get_height(b->gbm_bo),
+              1
+          },
+          .mipLevels = 1,
+          .arrayLayers = 1,
+          .samples = 1,
+          .tiling = VK_IMAGE_TILING_DRM_FORMAT_MODIFIER_EXT,
+          .usage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT,
+          .initialLayout = VK_IMAGE_LAYOUT_UNDEFINED,
+      };
+
+      res = vkCreateImage(vc->device, &base_create_info,
+                          NULL, &b->image);
+      fail_if(res != VK_SUCCESS, "failed to create image");
+
+      PFN_vkGetMemoryFdPropertiesKHR getMemoryFdPropertiesKHR =
+          (void *)vkGetDeviceProcAddr(vc->device, "vkGetMemoryFdPropertiesKHR");
+      fail_if(getMemoryFdPropertiesKHR == NULL,
+              "vkGetDeviceProcAddr(\"vkGetMemoryFdPropertiesKHR\") failed");
+
+      int prime_fd = gbm_bo_get_fd(b->gbm_bo);
```

Comment:
You already have `int fd; fd = gbm_bo_get_fd(b->gbm_bo);` above?

Reply:
fixed